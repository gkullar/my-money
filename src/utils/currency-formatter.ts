export const toGBP = (value: number): string => 'Â£' + (value / 100).toFixed(2);
