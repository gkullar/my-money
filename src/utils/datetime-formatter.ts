export const toDateTime = (value: string): string =>
  new Date(value).toLocaleString();
