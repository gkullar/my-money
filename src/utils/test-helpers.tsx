import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import 'jest-styled-components';
import AppProviders from '../app-providers';

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AppProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
