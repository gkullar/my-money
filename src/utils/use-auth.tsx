import React, {
  createContext,
  useContext,
  useState,
  FunctionComponent
} from 'react';
import uuid from './uuid';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;
const stateTokenKey = `${process.env.REACT_APP_API_STATE_TOKEN_KEY}`;
const clientId = `${process.env.REACT_APP_API_CLIENT_ID}`;
const clientSecret = `${process.env.REACT_APP_API_CLIENT_SECRET}`;
const redirectUri = 'http://localhost:3000'; // app not for production!

if (!localStorage.getItem(stateTokenKey))
  localStorage.setItem(stateTokenKey, uuid());

const stateToken = localStorage.getItem(stateTokenKey);

interface State {
  isAuthenticated: boolean;
  login: () => void;
  logout: () => void;
}

const authContext = createContext<State>({} as any);

const useAuth = () => useContext(authContext);

const AuthProvider: FunctionComponent<{}> = ({ children }) => {
  const auth = useAuthProvider();
  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
};

function useAuthProvider(): State {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(
    localStorage.getItem(accessTokenKey) ? true : false
  );
  const params = new URLSearchParams(window.location.search);

  if (!isAuthenticated && params.has('code') && params.has('state')) {
    if (params.get('state') !== stateToken)
      throw new Error('Error: state token not matched'); // @todo handle error in UI

    authenticate(params.get('code') as string);
  }

  async function authenticate(code: string) {
    const data = new FormData();
    data.append('grant_type', 'authorization_code');
    data.append('client_id', clientId);
    data.append('client_secret', clientSecret);
    data.append('redirect_uri', redirectUri);
    data.append('code', code);

    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/token`,
      {
        method: 'post',
        body: data
      }
    );

    const result = await response.json();

    if (response.status !== 200 || result.error)
      throw new Error('Error: Failed Authentication'); // @todo handle error in UI

    localStorage.setItem(accessTokenKey, result.access_token);
    setIsAuthenticated(true);
    window.history.replaceState(null, document.title, window.location.pathname); // remove query params
  }

  function login() {
    window.location.href = `https://auth.monzo.com/?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&state=${stateToken}`;
  }

  async function logout() {
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/logout`,
      {
        method: 'post',
        headers: new Headers({
          Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
        })
      }
    );

    const result = await response.json();

    if (!result.ok) throw new Error('Error: Failed to logout'); // @todo handle error in UI

    localStorage.removeItem(accessTokenKey);
    setIsAuthenticated(false);
  }

  return { isAuthenticated, login, logout };
}

export { useAuth, AuthProvider };
