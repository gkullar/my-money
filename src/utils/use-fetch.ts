import { useState, useEffect } from 'react';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;

interface State<T> {
  data: T;
  loading: boolean;
  hasErrors: boolean;
  hasPermissions: boolean;
}

const useFetch = <T>(url: string): State<T> => {
  const [data, setData] = useState<T>({} as T);
  const [loading, setLoading] = useState<boolean>(true);
  const [hasErrors, setHasErrors] = useState<boolean>(false);
  const [hasPermissions, setHasPermissions] = useState<boolean>(true);

  async function fetchUrl() {
    const response = await fetch(url, {
      method: 'get',
      headers: new Headers({
        Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
      })
    });

    const result = await response.json();

    if (response.status === 401) {
      localStorage.removeItem(accessTokenKey);
      window.location.href = '/';
    } else if (response.status === 403) {
      setHasPermissions(false);
    } else if (response.status !== 200) {
      throw new Error('Error: Failed Data Fetch'); // @todo handle error in UI
    } else {
      setData(result);
      setLoading(false);
      setHasErrors(false);
      setHasPermissions(true);
    }
  }

  useEffect(() => {
    fetchUrl();
  }, []);

  return { data, loading, hasErrors, hasPermissions };
};

export default useFetch;
