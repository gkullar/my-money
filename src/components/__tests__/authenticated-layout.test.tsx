import React from 'react';
import AuthenticatedLayout from '../authenticated-layout';
import { render } from '../../utils/test-helpers';

describe('AuthenticatedLayout', () => {
  test('renders correctly', () => {
    const { container } = render(<AuthenticatedLayout />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
