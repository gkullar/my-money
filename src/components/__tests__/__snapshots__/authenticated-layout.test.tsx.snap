// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticatedLayout renders correctly 1`] = `
.c10 {
  background: #fff;
  color: #1b1b1b;
  border-radius: 3px;
  border: 1px solid #37414f;
  padding: 5px 10px;
  text-transform: capitalize;
  -webkit-transition: background 0.3s linear;
  transition: background 0.3s linear;
}

.c10:hover {
  cursor: pointer;
  background: rgba(55,65,79,0.3);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  color: #1b1b1b;
  text-transform: capitalize;
  margin: 0 10px;
}

.c6 {
  display: inline-block;
  height: 20px;
  width: 40px;
  background: rgba(27,27,27,0.2);
  border-radius: 20px;
  cursor: pointer;
}

.c9 {
  height: 20px;
  width: 20px;
  border-radius: 20px;
  background: #fff;
  box-shadow: 0 1px 3px rgba(55,65,79,0.3);
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c7 {
  position: absolute;
  opacity: 0;
  cursor: pointer;
}

.c7:checked + .c8 {
  -webkit-transform: translate3d(100%,0,0);
  -ms-transform: translate3d(100%,0,0);
  transform: translate3d(100%,0,0);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100vh;
}

.c1 {
  position: relative;
  height: 60px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #1b1b1b;
  background: #fff;
  border-bottom: 1px solid rgba(55,65,79,0.3);
  padding: 5px 20px;
}

.c3 {
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 button {
  margin-left: 10px;
}

.c12 {
  color: #1b1b1b;
  background: #fff;
  border-top: 1px solid rgba(55,65,79,0.3);
  padding: 0 20px;
}

.c11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
  padding: 60px 20px 20px;
  height: 100%;
  background: #fff;
}

.c2 {
  position: absolute;
  top: 10px;
  height: 100px;
  background: #fff;
  -webkit-transition: position 0.4s linear,height 0.4s linear,top 0.4s linear;
  transition: position 0.4s linear,height 0.4s linear,top 0.4s linear;
}

<div
  class="c0"
>
  <header
    class="c1"
  >
    <img
      alt="logo"
      class="c2"
      src="logo.png"
    />
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          light
        </div>
        <label
          class="c6"
        >
          <input
            class="c7"
            type="checkbox"
          />
          <div
            class="c8 c9"
          />
        </label>
        <div
          class="c5"
        >
          dark
        </div>
      </div>
      <button
        class="c10"
      >
        logout
      </button>
    </div>
  </header>
  <main
    class="c11"
  />
  <footer
    class="c12"
  >
    <p>
      Environment: 
      test
    </p>
  </footer>
</div>
`;
